client/core/core.go

```diff
@@ -500,11 +506,75 @@ func (dc *dexConnection) runMatches(tradeMatches map[order.OrderID]*serverMatche
 // DRAFT NOTE: Right now, the matches are just checked and notifications sent,
 // but it may be a good place to trigger a FindRedemption if the conditions
 // warrant.
-func (dc *dexConnection) compareServerMatches(matches map[order.OrderID]*serverMatches) {
-	for _, match := range matches {
-		// readConnectMatches sends notifications for any problems encountered.
-		match.tracker.readConnectMatches(match.msgMatches)
+func (dc *dexConnection) compareServerMatches(srvMatches map[order.OrderID]*serverMatches) (exceptions map[order.OrderID]*tradeTrackers) {
+	exceptions = make(map[order.OrderID]*tradeTrackers)
+
+	// Identify extra matches named by the server response that we do not
+	// recognize.
+	for _, match := range srvMatches {
+		var extra []*msgjson.Match
+		match.tracker.mtx.RLock()
+		for _, msgMatch := range match.msgMatches {
+			var matchID order.MatchID
+			copy(matchID[:], msgMatch.MatchID)
+			if match.tracker.matches[matchID] == nil {
+				extra = append(extra, msgMatch)
+			}
+		}
+		match.tracker.mtx.RUnlock()
+		if len(extra) > 0 {
+			exceptions[match.tracker.ID()] = &tradeTrackers{
+				trade: match.tracker,
+				extra: extra,
+			}
+		}
+	}
+
+	in := func(matches []*msgjson.Match, mid []byte) bool {
+		for _, m := range matches {
+			if bytes.Equal(m.MatchID, mid) {
+				return true
+			}
+		}
+		return false
 	}
+
+	setMissing := func(trade *trackedTrade, missing []*matchTracker) {
+		if tt, found := exceptions[trade.ID()]; found {
+			tt.missing = missing
+		} else {
+			exceptions[trade.ID()] = &tradeTrackers{
+				trade:   trade,
+				missing: missing,
+			}
+		}
+	}
+
+	// Identify active matches that are missing from server's response.
+	dc.tradeMtx.RLock()
+	defer dc.tradeMtx.RUnlock()
+	for oid, trade := range dc.trades {
+		if !trade.isActive() {
+			continue // all of this trade's matches are inactive
+		}
+		tradeMatches, found := srvMatches[oid]
+		if !found {
+			// ALL of this trade's active matches are missing.
+			setMissing(trade, trade.activeMatches())
+			continue // check next local trade
+		}
+		// Check this local trade's active matches against server's reported
+		// matches for this trade.
+		var missing []*matchTracker
+		for _, match := range trade.activeMatches() { // each local match
+			if !in(tradeMatches.msgMatches, match.id[:]) { // against reported matches
+				missing = append(missing, match)
+			}
+		}
+		setMissing(trade, missing)
```

Comment:
Looks like you are creating an entry in `exceptions` even if `missing` is `nil`.